tools/memap.py

```diff
@@ -361,56 +479,136 @@ def parse_map_file_iar(self, file_desc):
-                        os.path.join(root, obj_file))
-
-                    if object_name in self.object_to_module:
-                        if DEBUG:
-                            print "WARNING: multiple usages of object file: %s"\
-                                % object_name
-                            print "    Current: %s" % \
-                                self.object_to_module[object_name]
-                            print "    New:     %s" % module_name
-                            print " "
-                    else:
-                        self.object_to_module.update({object_name:module_name})
+
+                    txt = os.path.join(root, obj_file)
+
+                    txt = txt.replace('\\', '/')
+
+                    # add relative path + object to list
+                    #self.dir_obj.append(txt[len(search_path)+1:])
+                    self.module_init(txt[len(search_path)+1:])
+
+
+    def reduce_depth(self, depth):
+        """
+        prints list of directories and objects. Examples:
+
+        (1) depth = 1:
+        main.o
+        mbed-os
+
+        (2) depth = 2:
+        main.o
+        mbed-os/test.o
+        mbed-os/drivers
+
+        """
+
+        # depth 0 shows all entries
+        if depth == 0:
+            self.short_modules = deepcopy(self.modules)
+            return
+
+        self.short_modules = dict()
+
+        # create reduced list
+        for line in self.modules:
+
+            data = line.split('/')
+            ndir = len(data)
+
+            temp = ''
+            count = 0
+
+            # iterate until the max depth level
+            max_level = min(depth, ndir)
+
+            # rebuild the path based on depth level
+            while count < max_level:
+                if count > 0:    # ignore '/' from first entry
+                    temp = temp + '/'
+
+                temp = temp + data[count]
+                count += 1
+
+            if temp not in self.short_modules:
+                temp_dic = dict()
+                for section_idx in self.all_sections:
+                    temp_dic[section_idx] = 0
+                self.short_modules[temp] = temp_dic
+
+            for section_idx in self.all_sections:
+                self.short_modules[temp][section_idx] += \
+                self.modules[line][section_idx]
+
 
     export_formats = ["json", "csv-ci", "table"]
 
-    def generate_output(self, export_format, file_output=None):
+    def generate_output(self, export_format, depth, file_output=None):
         """ Generates summary of memory map data
 
         Positional arguments:
         export_format - the format to dump
 
         Keyword arguments:
         file_desc - descriptor (either stdout or file)
+        depth - directory depth on report
 
         Returns: generated string for the 'table' format, otherwise None
         """
 
+        self.module_remove_unused() # clean up unused modules/objects
```

Comment:
I think this method is named well enough that we don't need this comment.

Not blocking this PR.